src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -657,8 +673,134 @@ public override Task WriteAsync(Byte[] array, int offset, int count, Cancellatio
 
         public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
         {
-            return StreamHelpers.ArrayPoolCopyToAsync(this, destination, bufferSize, cancellationToken: cancellationToken);
+            // Validation as base CopyToAsync would do
+            StreamHelpers.ValidateCopyToArgs(this, destination, bufferSize);
+
+            // Validation as ReadAsync would do
+            EnsureDecompressionMode();
+            if (_asyncOperations != 0) throw new InvalidOperationException(SR.InvalidBeginCall);
+            EnsureNotDisposed();
+
+            // Early check for cancellation
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled<int>(cancellationToken);
+            }
+
+            // The caller asked for a specific bufferSize, but we're going to end up potentially using two buffers:
+            // a buffer for storing the inflated output and passing along to the destination stream, and whatever
+            // buffer the source stream creates in its CopyToAsync implementation.  As such, we halve the buffer size,
+            // to split it across the two.
+            bufferSize = Math.Max(1, bufferSize / 2);
```

Comment:
Fair question.  My primary concern is around the case where the buffer pool is exhausted.  If there are plenty of available buffers in the pool, then using 2 instead of 1 at the requested size isn't that big a deal.  But if there aren't any buffers in the pool, we're going to now end up allocating 2 buffers where we previously would have only allocated 1.  I figured we could at least allocate no more in total size than the user requested in such cases.

I don't have enough data to really argue one way or the other, though.  If you think I should remove this and just go with the original bufferSize, I can.